### DISCUSSION TITLE
Reading Object type Resource from Device Service

### DISCUSSION THREAD

**Discussion by sounak-1989** [edgexfoundry/community#25] [General]
I wanted to understand how I can read a Resource of Type "Object" from the device service. When I send the structure the "objectValue" field is just blank.

Do I need to convert the structure into a JSON object before sending? The Virtual device example does not have an example for it. It will be great If there is an example or explanation on how to do it.

Thanks.

**Comment by cloudxxx8** [comment]
do you mean **write** Object instead of **read**?  You can read Object from virtual device service via the GetCommand.
EdgeX doesn't support write Object data type (SetCommand) so far.

**Reply by sounak-1989** [reply]
For example if I define a resource in my device profile as Object type and It has "Read" permissions only. How should I arrange the structure so that it is showed in the HTTP response body in proper format.

**Comment by sounak-1989** [comment]
Hey guys figured it out. For sending objects as response to read handler it should be of the form map[string]interface{}.
To align my structure to the above structure I converted mine to a json struct and then marshalled and unmarshalled the data into the above structure.

Let me know that's a good way of doing it or is there a better way.
Thanks

**Reply by cloudxxx8** [reply]
thanks for the update.  sorry, I was confused between "Object" and "Binary" value type.  Yes, you are using the write way to parse the Object value unless you have a matched struct defined in your code.

**Comment by sounak-1989** [comment]
The following is the code that does the conversion.
func (s *SimpleDriver) GetDout(deviceName string, portId string) map[string]interface{} {
	var resp map[string]interface{}
	var digOut DigitalOutReading
	// Get Device Object from deviceName
	reqId, _ := strconv.Atoi(s.activeDevices[deviceName].deviceInfo.deviceID)
	reqDevice := s.spiSubsystem.GetDeviceMap(true)[uint32(reqId)]
	portIdInt, _ := strconv.Atoi(portId)
	// Get current Settings byte array
	reading := reqDevice.GetReadings()[(portIdInt * 4) : (portIdInt*4)+4]
	buf := bytes.NewReader(reading)
	_ = binary.Read(buf, binary.LittleEndian, &digOut.outputState)
	_ = binary.Read(buf, binary.LittleEndian, &digOut.mode)
	_ = binary.Read(buf, binary.LittleEndian, &digOut.dummy)
	var outState string
	if digOut.outputState == 1 {
		outState = "ON"
	} else {
		outState = "OFF"
	}
	digOutJson := &DigitalOutReadingJson{
		OpState: outState,
		Mode:    strconv.Itoa(int(digOut.mode)),
	}
	data, err := json.Marshal(digOutJson)
	fmt.Println(string(data))
	if err != nil {
		fmt.Println(err.Error())
	}
	err = json.Unmarshal(data, &resp)
	if err != nil {
		fmt.Println(err.Error())
	}
	return resp
}
